File: src/interrupts/pic.c
#include "interrupts/pic.h"
#include "io/io.h"
#include "debug/debug.h"

// PIC ports
#define PIC1_COMMAND 0x20
#define PIC1_DATA    0x21
#define PIC2_COMMAND 0xA0
#define PIC2_DATA    0xA1

// PIC commands
#define PIC_EOI      0x20
#define ICW1_ICW4    0x01
#define ICW1_SINGLE  0x02
#define ICW1_INTERVAL4 0x04
#define ICW1_LEVEL   0x08
#define ICW1_INIT    0x10
#define ICW4_8086    0x01

// IRQ mappings
#define IRQ_BASE     0x20
#define IRQ_COUNT    16

/**
 * Send End-of-Interrupt to PIC
 */
void pic_send_eoi(uint8_t irq) {
    if (irq >= 8) {
        port_write_byte(PIC2_COMMAND, PIC_EOI);
    }
    port_write_byte(PIC1_COMMAND, PIC_EOI);
}

/**
 * Mask (disable) an IRQ
 */
void pic_mask_irq(uint8_t irq) {
    uint16_t port;
    uint8_t value;
    
    if (irq < 8) {
        port = PIC1_DATA;
    } else {
        port = PIC2_DATA;
        irq -= 8;
    }
    
    value = port_read_byte(port) | (1 << irq);
    port_write_byte(port, value);
}

/**
 * Unmask (enable) an IRQ
 */
void pic_unmask_irq(uint8_t irq) {
    uint16_t port;
    uint8_t value;
    
    if (irq < 8) {
        port = PIC1_DATA;
    } else {
        port = PIC2_DATA;
        irq -= 8;
    }
    
    value = port_read_byte(port) & ~(1 << irq);
    port_write_byte(port, value);
}

/**
 * Initialize the PIC
 */
void pic_init(void) {
    DEBUG_INFO("PIC", "Initializing PIC");
    
    uint8_t mask1, mask2;
    
    // Save masks
    mask1 = port_read_byte(PIC1_DATA);
    mask2 = port_read_byte(PIC2_DATA);
    
    DEBUG_INFO("PIC", "Remapping PIC");
    
    // Start initialization sequence
    port_write_byte(PIC1_COMMAND, ICW1_INIT | ICW1_ICW4);
    io_wait();
    port_write_byte(PIC2_COMMAND, ICW1_INIT | ICW1_ICW4);
    io_wait();
    
    // Set vector offsets
    port_write_byte(PIC1_DATA, IRQ_BASE);     // IRQ 0-7: int 0x20-0x27
    io_wait();
    port_write_byte(PIC2_DATA, IRQ_BASE + 8); // IRQ 8-15: int 0x28-0x2F
    io_wait();
    
    // Tell PICs about each other
    port_write_byte(PIC1_DATA, 4);            // Tell Master about Slave at IRQ2
    io_wait();
    port_write_byte(PIC2_DATA, 2);            // Tell Slave its cascade identity
    io_wait();
    
    // Set 8086 mode
    port_write_byte(PIC1_DATA, ICW4_8086);
    io_wait();
    port_write_byte(PIC2_DATA, ICW4_8086);
    io_wait();
    
    DEBUG_INFO("PIC", "Setting interrupt masks");
    
    // Mask all interrupts except:
    // - IRQ0 (timer)
    // - IRQ1 (keyboard)
    // - IRQ2 (cascade)
    // - IRQ12 (PS/2 mouse)
    mask1 = ~((1 << 0) | (1 << 1) | (1 << 2));  // Enable IRQ0, IRQ1, IRQ2
    mask2 = ~(1 << 4);  // Enable IRQ12 (12-8 = 4 in the second PIC)
    
    // Set masks
    port_write_byte(PIC1_DATA, mask1);
    port_write_byte(PIC2_DATA, mask2);
    
    DEBUG_INFO("PIC", "PIC initialized");
}

/**
 * Disable the PIC (useful for APIC mode)
 */
void pic_disable(void) {
    DEBUG_INFO("PIC", "Disabling PIC");
    port_write_byte(PIC1_DATA, 0xFF);
    port_write_byte(PIC2_DATA, 0xFF);
} 

File: kernel/src/interrupts/pic.rs
// kernel/src/interrupts/pic.rs
//! PIC (Programmable Interrupt Controller) management
//! This module provides a clean interface for initializing and configuring the 8259A PICs.

use spin::Mutex;
use crate::serial_println;
use x86_64::instructions::port::Port;

/// The PICs are configured to use these offsets for their IRQs
pub const PIC_1_OFFSET: u8 = 32;
pub const PIC_2_OFFSET: u8 = PIC_1_OFFSET + 8;

/// Command and data ports for the PICs
const PIC_1_COMMAND: u16 = 0x20;
const PIC_1_DATA: u16 = 0x21;
const PIC_2_COMMAND: u16 = 0xA0;
const PIC_2_DATA: u16 = 0xA1;

/// PIC initialization command words
const ICW1_INIT: u8 = 0x10;
const ICW1_ICW4: u8 = 0x01;
const ICW4_8086: u8 = 0x01;

/// PIC end of interrupt command
const PIC_EOI: u8 = 0x20;

/// Represents interrupt vectors corresponding to PIC IRQs
#[derive(Debug, Clone, Copy)]
#[repr(u8)]
pub enum InterruptIndex {
    Timer = PIC_1_OFFSET,      // IRQ 0
    Keyboard = PIC_1_OFFSET + 1, // IRQ 1
    // IRQs 2-7 on the master PIC
    
    // IRQs 8-15 on the slave PIC, add 8 to the offset
    Mouse = PIC_1_OFFSET + 12, // IRQ 12
    // Add other PIC interrupts as needed
}

impl InterruptIndex {
    /// Convert to the raw u8 interrupt vector
    pub fn as_u8(self) -> u8 {
        self as u8
    }

    /// Convert to usize for IDT indexing
    pub fn as_usize(self) -> usize {
        usize::from(self.as_u8())
    }
}

/// A structure representing the PIC controller
pub struct PicController {
    primary_offset: u8,
    secondary_offset: u8,
}

impl PicController {
    /// Creates a new PIC controller with the given offsets
    pub const fn new(primary_offset: u8, secondary_offset: u8) -> Self {
        Self {
            primary_offset,
            secondary_offset,
        }
    }

    /// Initializes the PICs with the configured offsets
    pub fn initialize(&mut self) {
        serial_println!("Initializing PICs with offsets: primary={}, secondary={}", 
            self.primary_offset, self.secondary_offset);

        // Save current masks
        let primary_mask: u8 = unsafe { Port::new(PIC_1_DATA).read() };
        let secondary_mask: u8 = unsafe { Port::new(PIC_2_DATA).read() };
        serial_println!("DEBUG: Saved PIC masks - Primary: {:08b}, Secondary: {:08b}", 
            primary_mask, secondary_mask);

        // Start initialization sequence
        unsafe {
            // ICW1: Start initialization sequence
            Port::new(PIC_1_COMMAND).write(ICW1_INIT | ICW1_ICW4);
            Port::new(PIC_2_COMMAND).write(ICW1_INIT | ICW1_ICW4);

            // ICW2: Set vector offsets
            Port::new(PIC_1_DATA).write(self.primary_offset);
            Port::new(PIC_2_DATA).write(self.secondary_offset);

            // ICW3: Tell PICs how they're cascaded
            Port::new(PIC_1_DATA).write(4u8); // Secondary PIC at IRQ2
            Port::new(PIC_2_DATA).write(2u8); // Cascade identity

            // ICW4: Set 8086 mode
            Port::new(PIC_1_DATA).write(ICW4_8086);
            Port::new(PIC_2_DATA).write(ICW4_8086);

            // Mask all interrupts initially
            Port::new(PIC_1_DATA).write(0xFFu8);
            Port::new(PIC_2_DATA).write(0xFFu8);
        }

        serial_println!("PIC initialization complete");
    }

    /// Configures which IRQs are enabled/disabled
    pub fn configure_irqs(&mut self, primary_mask: u8, secondary_mask: u8) {
        serial_println!("Configuring IRQs - Primary mask: {:08b}, Secondary mask: {:08b}", 
            primary_mask, secondary_mask);
            
        unsafe {
            // Ensure we're not enabling any interrupts that aren't properly set up
            let safe_primary_mask = primary_mask & 0xFCu8; // Only allow IRQ0 (timer) and IRQ1 (keyboard)
            let safe_secondary_mask = secondary_mask & 0xEFu8; // Only allow IRQ12 (mouse)
            
            // Write the masks
            Port::new(PIC_1_DATA).write(safe_primary_mask);
            Port::new(PIC_2_DATA).write(safe_secondary_mask);
            
            // Verify the masks were written correctly
            let verify_primary: u8 = unsafe { Port::new(PIC_1_DATA).read() };
            let verify_secondary: u8 = unsafe { Port::new(PIC_2_DATA).read() };
            
            serial_println!("IRQs configured - Primary: {:08b}, Secondary: {:08b}", 
                safe_primary_mask, safe_secondary_mask);
            serial_println!("Verified masks - Primary: {:08b}, Secondary: {:08b}", 
                verify_primary, verify_secondary);
        }
    }

    /// Sends an end of interrupt signal for the given IRQ
    pub fn notify_end_of_interrupt(&mut self, irq: u8) {
        if irq >= 8 {
            unsafe {
                Port::new(PIC_2_COMMAND).write(PIC_EOI);
            }
        }
        unsafe {
            Port::new(PIC_1_COMMAND).write(PIC_EOI);
        }
    }
}

/// Global PIC controller instance
pub static PIC_CONTROLLER: Mutex<PicController> = Mutex::new(PicController::new(PIC_1_OFFSET, PIC_2_OFFSET)); 